### PR TITLE
HA Feature delivery

### PR DESCRIPTION
**Pull request by chiraggupta06**
#### What does this PR do?
HA feature 
#### Do you have any concerns with this PR?
NA
#### How can the reviewer verify this PR?
Deploying k8s with HA 
#### Any background context you want to provide?
NA
#### Screenshots or logs (if appropriate)
NA
#### Questions:
- Have you connected this PR to the issue it resolves?
- Does the documentation need an update?
- Does this add new Python dependencies?
- Have you added unit or functional tests for this PR?
- Does this patch update any configuration files?

### CODE REVIEW

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+# This script is responsible for deploying Aricent_Iaas environments and
+# Kubernetes Services
+--- 
+- 
+  become: true
+  become_user: root
+  hosts: "{{ host_name }}"
+  vars_files:
+    - "{{ VARIABLE_FILE }}"
```
Comment (spisarski) [inline snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: These files no longer exist. You will need to pass in all parameters via the ansible_utils API call

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      command: service haproxy status
+      register: haproxy_status
+      failed_when: "'inactive (dead)' in haproxy_status.stdout"
+    -
+      debug: "msg={{ haproxy_status.stdout }}"
+
+    -
+      name: check file existence of haproxy.cfg in   {{ PROJECT_PATH }}{{Project_name}}
+      register: projcet_haproxy_cfg
+      stat: path={{ PROJECT_PATH }}{{Project_name}}/haproxy.cfg
```
Comment (spisarski) [inline snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: Please use config_utils.get_project_artifact_dir() to obtain the {{ PROJECT_PATH }}{{ PROJECT_NAME }} and replace with {{ PROJ_ARTIFACT_DIR }}

File: snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is responsible for deploying Aricent_Iaas environments and
+# Kubernetes Services
+--- 
+- 
+  become: true
+  hosts: "{{ ip }}"
```
Comment (spisarski) [inline snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml]: Please do not send in the IP here and use something like 'all' in this case and send in a list of IPs into the ansible_utils API call

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is responsible for deploying Aricent_Iaas environments and
+# Kubernetes Services
+--- 
+- 
+  become: true
+  become_user: root
+  hosts: "{{ host_name }}"
```
Comment (spisarski) [inline snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: See comment above over the hosts value.

File: snaps_k8s/provision/ansible_configuration.py [hunk 5]
```diff
@@ -232,7 +232,27 @@ def __kubespray(k8s_conf, base_pb_vars):
...
+        logger.info("INSIDE KUBESPPRAY Load balancer ip %s", lb_ip)
+        ha_enabled = True
+        pb_vars = {
+        'SRC_PACKAGE_PATH': config_utils.get_artifact_dir(k8s_conf),
+        'PROJ_ARTIFACT_DIR': config_utils.get_project_artifact_dir(k8s_conf),
+        'Project_name': config_utils.get_project_name(k8s_conf),
+        }
 
+        pb_vars.update(config_utils.get_proxy_dict(k8s_conf))
+        loadbalancer_dict = __get_loadbalancer_dict(k8s_conf)
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Can you move __get_loadbalancer_dict() into config_utils?

File: snaps_k8s/provision/ansible_configuration.py [hunk 6]
```diff
@@ -232,7 +232,27 @@ def __kubespray(k8s_conf, base_pb_vars):
...
+        if loadbalancer_dict:
+            __ha_configuration(loadbalancer_dict,pb_vars,k8s_conf,config_utils.get_project_name(k8s_conf))
+
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Why not just send in the k8s_conf here and extract the loadbalancer_dict, pb_vars, and project name inside of the __ha_configuration() function?

File: snaps_k8s/provision/ansible_configuration.py [hunk 7]
```diff
@@ -758,6 +778,13 @@ def __install_kubectl(k8s_conf):
 
     host_name, ip = config_utils.get_first_master_host(k8s_conf)
     ha_enabled = len(lb_ips) > 0
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please get the HA information from config_utils

File: snaps_k8s/provision/ansible_configuration.py [hunk 8]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
         }
         ansible_utils.apply_playbook(consts.K8_DELETE_WEAVE_INTERFACE,
                                      variables=pb_vars)
+def __get_loadbalancer_dict(config):
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please move me to config_utils.py or change the get_ha_config() function to match as other calls would not  be really working anyhow

File: snaps_k8s/provision/ansible_configuration.py [hunk 9]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+    function to get loadbalancer dictionary
+    :param config:
+    :return:
+    """
+    logger.info("Argument List:\n config: %s", config)
+    for item in config_utils.get_ha_config(config):
+        return item.get(consts.HA_API_EXT_LB_KEY)
+                        
+
+def __get_hostname_map(hosts):
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please leverage config_utils#get_hostname_ips_dict() or get_nodes_ip_name_type() and remove this function.

File: snaps_k8s/provision/ansible_configuration.py [hunk 10]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+            hostname = hosts[i].get(consts.HOST_KEY).get('hostname')
+            host_ip = ""
+            ip_addr = hosts[i].get(consts.HOST_KEY).get(consts.IP_KEY)
+            if ip_addr:
+                host_ip = ip_addr
+            hostname_map[hostname] = host_ip
+
+    logger.info('Exit')
+    return hostname_map
+def __create_host_nodetype_map(hosts):
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please leverage config_utils#get_hostname_ips_dict() or get_nodes_ip_name_type() and remove this function.

File: snaps_k8s/provision/ansible_configuration.py [hunk 11]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+    logger.info('HA CONFIGURING')
+    ret_val_lb_config = __launch_ha_loadbalancer_conf(
+        hostname_map=hostname_map, host_node_type_map=host_node_type_map,
+        loadbalancer_dict=loadbalancer_dict, project_name=project_name,pb_vars=pb_vars)
+    ret_val_ha_config = __launch_kubespray_ha_configure(
+        hostname_map=hostname_map, host_node_type_map=host_node_type_map,
+        loadbalancer_dict=loadbalancer_dict, project_name=project_name,pb_vars=pb_vars)
+    if not(ret_val_lb_config and ret_val_ha_config):
+        logger.error('Alert !! Failed in HA Configuration')
+        exit(1)
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please do not add explicit exit() calls. Raised exceptions should bubble all of the way up to main unless expected.

File: snaps_k8s/provision/ansible_configuration.py [hunk 12]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+    try:
+        ret_val = __launch_load_balancer(
+            consts.K8_HA_EXT_LB,
+            loadbalancer_dict.get(consts.IP_KEY),
+            loadbalancer_dict.get(consts.HOSTNAME_KEY),
+            master_ip_list, lb_port,
+            loadbalancer_ip=loadbalancer_dict.get(consts.IP_KEY),
+            project_name=project_name,
+            base_pb_vars_ha=pb_vars)
+    except Exception as exception:
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Please do not catch exception

File: snaps_k8s/provision/ansible_configuration.py [hunk 13]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+        logger.error(exception.message)
+        logger.info("exception while calling "
+                    "external_load_balancer_install ")
+        ret_val = False
+    return ret_val
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: No need for a return value as exceptions should not be caught ever when fatal

File: snaps_k8s/provision/ansible_configuration.py [hunk 14]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+    ret_val = False
+    logger.info("launch_ha_kubespray_configure function")
+    host_node_type_map_local = {'localhost': '127.0.0.1'}
+    try:
+        ret_val = __kubespray_ha_configure(
+            consts.K8_HA_KUBESPRAY_CONFIGURE,
+            host_node_type_map_local['localhost'], 'localhost',
+            loadbalancer_dict.get(consts.IP_KEY),
+            project_name=project_name,pb_vars=pb_vars)
+    except Exception as exception:
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Remove try/except block

File: snaps_k8s/provision/ansible_configuration.py [hunk 15]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+        logger.error(exception.message)
+        logger.info("exception while calling launch_ha_kubespray_configure")
+        ret_val = False
+    return ret_val
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Do not return a value

File: snaps_k8s/provision/ansible_configuration.py [hunk 16]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+    logger.info("Argument List:\n hostname_map: %s\n host_node_type_map: "
+                "%s\n loadbalancer_dict: %s\n project_name: %s",
+                hostname_map, host_node_type_map, loadbalancer_dict,
+                project_name)
+
+    ret_val = False
+    logger.info("launch_ha_kubespray_configure function")
+    host_node_type_map_local = {'localhost': '127.0.0.1'}
+    try:
+        ret_val = __kubespray_ha_configure(
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Do not store return value. Exception handling will deal with this properly

File: snaps_k8s/provision/ansible_configuration.py [hunk 17]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+        'host_name': host_name,
+        'MASTER_IP_LIST': master_ip_list,
+        'lb_port': lb_port,
+        'LOADBALANCER_IP': loadbalancer_ip,
+    }
+    pb_vars.update(base_pb_vars_ha)
+    logger.info("Arguments supali are %s", pb_vars)
+  
+    ansible_utils.apply_playbook(playbook ,[loadbalancer_ip],variables=pb_vars) 
+    return True
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Remove return. Exceptions raise by apply_playbook will handle things

File: snaps_k8s/provision/ansible_configuration.py [hunk 18]
```diff
@@ -902,3 +929,189 @@ def delete_weave_interface(k8s_conf):
...
+        'ip': ip_addr,
+        'host_name': host_name,
+        'LOADBALANCER_IP': loadbalancer_ip,
+        'Project_name': project_name,
+    }
+    pb_vars_1.update(pb_vars)
+    logger.info("Arguments are %s", pb_vars_1)
+    ansible_utils.apply_playbook(playbook ,variables=pb_vars_1)
+    logger.info('Exit')
+    return True
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Remove return. Exceptions raise by apply_playbook will handle things

File: snaps_k8s/provision/ansible_configuration.py [hunk 19]
```diff
@@ -232,7 +232,17 @@ def __kubespray(k8s_conf, base_pb_vars):
             consts.K8_CPU_PINNING_CONFIG,
             variables={'KUBESPRAY_PATH': config_utils.get_kubespray_dir(
                 k8s_conf)})
-
+                
+    ha_enabled = False
+    ha_configuration = config_utils.get_ha_config(k8s_conf)
+    if ha_configuration:
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: This is ok but I'd like to see a function in config_utils that will simply return the boolean for ha_enabled.

File: snaps_k8s/provision/ansible_configuration.py [hunk 20]
```diff
@@ -758,6 +767,13 @@ def __install_kubectl(k8s_conf):
 
     host_name, ip = config_utils.get_first_master_host(k8s_conf)
     ha_enabled = len(lb_ips) > 0
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: Why is ha_enabled being set again? It was set above but never used.

**Comment by spisarski** [comment]
@chiraggupta06 - Please also integrate the HA CI scripts that can be found in the branch 'ha_ci_scripts'

File: snaps_k8s/provision/ansible_configuration.py (hunk 7 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Hi Steve,

value of  ha_enabled is being fetched from ib_ip which is being extracted from config_utils file.
code snippet-
 lb_ip = "127.0.0.1"
    lb_ips = config_utils.get_ha_lb_ips(k8s_conf)
    if len(lb_ips) > 0:
        lb_ip = lb_ips[0]

    logger.info("Load balancer ip %s", lb_ip)

    host_name, ip = config_utils.get_first_master_host(k8s_conf)
    ha_enabled = len(lb_ips) > 0

Please let me know if you have any input.

File: snaps_k8s/provision/ansible_configuration.py (hunk 16 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml (hunk 1 above)
Comment (chiraggupta06) [reply snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: Incorporated.

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml (hunk 2 above)
Comment (chiraggupta06) [reply snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: Incorporated.

File: snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml (hunk 3 above)
Comment (chiraggupta06) [reply snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml]: Incorporated.

File: snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml (hunk 4 above)
Comment (chiraggupta06) [reply snaps_k8s/playbooks/k8/k8_ha_multicluster_loadbalancer_configure.yaml]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 5 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 6 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 14 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 15 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 17 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 18 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: Incorporated.

File: snaps_k8s/provision/ansible_configuration.py (hunk 19 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: comments incorporated. Removed ha_enabled flag as it is no longer used after comments incorporation.

File: snaps_k8s/provision/ansible_configuration.py (hunk 20 above)
Comment (chiraggupta06) [reply snaps_k8s/provision/ansible_configuration.py]: comment incorporated.  removed other ha_enabled flag. now this is the only flag that is being used.

File: snaps_k8s/provision/ansible_configuration.py [hunk 21]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+    pb_vars = {
+        'ip': ip_addr,
+        'host_name': host_name,
+        'MASTER_IP_LIST': master_ip_list,
+        'lb_port': lb_port,
+        'LOADBALANCER_IP': loadbalancer_ip,
+    }
+    pb_vars.update(base_pb_vars_ha)
+    logger.info("Arguments are %s", pb_vars)
+    ansible_utils.apply_playbook(playbook, [loadbalancer_ip], variables=pb_vars)
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: You need to add the user here.
    ansible_utils.apply_playbook(
        playbook, [loadbalancer_ip], consts.NODE_USER, variables=pb_vars)

File: snaps_k8s/provision/ansible_configuration.py [hunk 22]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+    """
+    logger.info("Argument List:\n hostname_map: %s\n host_node_type_map: "
+                "%s\n loadbalancer_dict: %s \n pb_vars: %s",
+                hostname_map, host_node_type_map, loadbalancer_dict,
+                pb_vars)
+
+    logger.info("launch_ha_loadbalancer_conf function")
+    host_node_type_map_local = {}
+    lb_port = str(loadbalancer_dict.get("port"))
+    master_ip_list = []
```
Comment (spisarski) [inline snaps_k8s/provision/ansible_configuration.py]: use config_utils.get_master_node_ips(k8s_config) to obtain this master list; however, this logic doesn't seem correct as it should be the lb_node. When running the HA CI script, it appears to have picked up the correct one though. Please either use the config_utils function to obtain the master IPs or explain this logic to me.

File: snaps_k8s/playbooks/k8/k8_ha_kubespray_configure.yaml [hunk 23]
```diff
@@ -14,5 +14,34 @@
 # limitations under the License.
 ###########################################################################
 
+--- 
+- 
+  hosts: localhost
+  become: true
+  become_user: root
```
Comment (spisarski) [inline snaps_k8s/playbooks/k8/k8_ha_kubespray_configure.yaml]: needs a privilege escalation method here:
  become_method: sudo

**Comment by spisarski** [comment]
Got the 'ha_ci_scripts' branch working after rebasing with your enhancements.

File: snaps_k8s/provision/ansible_configuration.py (hunk 9 above)
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: I have implemented in the 'ha_ci_scripts' branch

File: snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is responsible for deploying Aricent_Iaas environments and
+# Kubernetes Services
+--- 
+- 
+  #become: true
```
[the remark this answers is not in the document]
Comment (spisarski) [reply snaps_k8s/playbooks/k8/k8_ha_external_load_balancer_install.yaml]: removed in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py (hunk 21 above)
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: fixed in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py (hunk 9 above)
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: implemented in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py (hunk 22 above)
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: implemented in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py [hunk 25]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+                             loadbalancer_ip, pb_vars=None):
+    """
+    function added for ha configuration in kubespray
+    :param playbook:
+    :param ip_addr:
+    :param host_name:
+    :param loadbalancer_ip:
+    """
+    pb_vars_1 = {
+        'ip': ip_addr,
```
[the remark this answers is not in the document]
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: fixed in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py [hunk 26]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+    """
+    function added for ha install load balancer
+    :param playbook:
+    :param ip_addr:
+    :param host_name:
+    :param master_ip_list:
+    :param lb_port:
+    :param loadbalancer_ip:
+    """
+    pb_vars = {
```
[the remark this answers is not in the document]
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: changed in  in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py [hunk 27]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+    logger.info("launch_ha_kubespray_configure function")
+    host_node_type_map_local = {'localhost': '127.0.0.1'}
+    __kubespray_ha_configure(
+        consts.K8_HA_KUBESPRAY_CONFIGURE,
+        host_node_type_map_local['localhost'], 'localhost',
+        loadbalancer_dict.get(consts.IP_KEY),
+        pb_vars=pb_vars)
+
+
+def __launch_load_balancer(playbook, ip_addr, host_name,
```
[the remark this answers is not in the document]
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: changed  in 'ha_ci_scripts' branch

File: snaps_k8s/provision/ansible_configuration.py [hunk 28]
```diff
@@ -902,3 +913,159 @@ def delete_weave_interface(k8s_conf):
...
+        'MASTER_IP_LIST': master_ip_list,
+        'lb_port': lb_port,
+        'LOADBALANCER_IP': loadbalancer_ip,
+    }
+    pb_vars.update(base_pb_vars_ha)
+    logger.info("Arguments are %s", pb_vars)
+    ansible_utils.apply_playbook(playbook, [loadbalancer_ip], variables=pb_vars)
+
+
+def __kubespray_ha_configure(playbook, ip_addr, host_name,
```
[the remark this answers is not in the document]
Comment (spisarski) [reply snaps_k8s/provision/ansible_configuration.py]: changed in 'ha_ci_scripts' branch

File: snaps_k8s/playbooks/k8/k8_ha_kubespray_configure.yaml (hunk 23 above)
Comment (spisarski) [reply snaps_k8s/playbooks/k8/k8_ha_kubespray_configure.yaml]: fixed in 'ha_ci_scripts' branch

**Comment by spisarski** [comment]
Please validate the HA environment at ubuntu@10.197.113.109 pass 'Pa$$w0rd' and the non-HA environment spun up with the same 'ha_ci_scripts' branch rebased with your PR + code review comments implemented at ubuntu@10.197.113.121 'Pa$$w0rd'

**Pull request merged**